ansible/roles/test/files/ptftests/sonic.py

```diff
@@ -0,0 +1,618 @@
+        parsed_logs = self.extract_from_logs(syslog_regex_r, log_lines, min_timestamp=start_time)
+        self.log('Log output "{}"'.format('\n'.join(["{} {} {}".format(k[0], j, k[1]) for j in parsed_logs for k in parsed_logs[j]])))
+        log_data = self.parse_logs(parsed_logs)
+        if (self.reboot_type == 'fast-reboot' and \
+            'bgp#bgpd' in parsed_logs and 'PortChannel' in parsed_logs) \
+                or (self.reboot_type == 'warm-reboot' and 'bgp#bgpd' in parsed_logs) \
+                or (self.reboot_type == 'service-warm-restart' and 'bgp#bgpd' in parsed_logs):
+            log_present = True
+        else:
+            log_data['error'] = 'Incomplete logs'
+
+        self.log('Disconnecting from VM')
+        self.disconnect()
+
+        # save data for troubleshooting
+        with open("/tmp/%s.data.pickle" % self.ip, "w") as fp:
+            pickle.dump(data, fp)
+
+        # save debug data for troubleshooting
+        if self.DEBUG:
+            with open("/tmp/%s.raw.pickle" % self.ip, "w") as fp:
+                pickle.dump(debug_data, fp)
+            with open("/tmp/%s.logging" % self.ip, "w") as fp:
+                fp.write("\n".join(log_lines))
+
+        self.log('Checking BGP GR peer status on VM')
+        self.check_gr_peer_status(data)
+        cli_data = {}
+        #cli_data['lacp']   = self.check_series_status(data, "lacp",         "LACP session")
+        cli_data['lacp']   = (0, 0)
+        cli_data['bgp_v4'] = self.check_series_status(data, "bgp_route_v4", "BGP v4 routes")
+        cli_data['bgp_v6'] = self.check_series_status(data, "bgp_route_v6", "BGP v6 routes")
+        cli_data['po']     = self.check_lag_flaps("PortChannel1", log_lines, start_time)
+
+        if 'route_timeout' in log_data:
+            route_timeout             = log_data['route_timeout']
+            cli_data['route_timeout'] = route_timeout
+
+            # {'10.0.0.38': [(0, '4200065100)')], 'fc00::2d': [(0, '4200065100)')]}
+            for nei in route_timeout.keys():
+                asn = route_timeout[nei][0][-1]
+                msg = 'BGP route GR timeout: neighbor %s (ASN %s' % (nei, asn)
+                self.fails.add(msg)
+
+        self.log('Finishing run()')
+        return self.fails, self.info, cli_data, log_data, {"lacp_down": self.lacp_pdu_time_on_down, "lacp_up": self.lacp_pdu_time_on_up}
+
+    def extract_from_logs(self, regexp, data, min_timestamp=None):
+        raw_data = []
+        result = defaultdict(list)
+        re_compiled = re.compile(regexp)
+        for line in data:
+            m = re_compiled.match(line)
+            if not m:
+                continue
+            log_time = datetime.datetime.strptime(str(datetime.datetime.now().year) + " " + m.group(1), "%Y %b %d %X")
+            # Python 3 version (Python 2 doesn't have timestamp():
+            #raw_data.append((log_time.timestamp(), m.group(2), m.group(3)))
```

Comment:
We need to support testing on both Python2 and 3.

Reply:
That line will work for both Python 2 and 3; if we do move to Python 3-only, then the next line could be used for conciseness (although looking at it now, that Python 3 version needs to be updated to match the current code).